homeassistant/components/modbus/__init__.py

```diff
@@ -150,8 +149,8 @@ def start_modbus(event):
         client_name = service.data[ATTR_HUB]
         await hub_collect[client_name].write_coil(unit, address, state)
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_modbus)
-
+    # do not wait for EVENT_HOMEASSISTANT_START, activate pymodbus now
+    start_modbus()
```

Comment:
Is this call async safe?

We're mixing sync and async api inside `start_modbus`. We should select either of those APIs depending on where it should run, inside event loop or outside.

Reply:
yeah I am painfully aware of that. Problem is that pymodbus needs a sync setup call, while all of our code is async. I will isolate the setup call in sync shield.

Reply:
We can use the executor if it should run in another thread. Use `hass.async_add_executor_job` for that.

Reply:
I am trying exactly that, but for some reason the setup hangs, so obviously I have done something wrong.

Reply:
Push what you have and I can take a look.

Reply:
thanks. I have pushed my latest.

I think the old code was attaching to the start event to secure a pure sync call (at least I remember having seen this problem earlier, when I converted everything to async).

I

Reply:
“madre mia” ! (a good spanish expression. It was hanging in listen_once, and I also needed an await (of course) in the executor line.

Now it works for me.